docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,438 @@
+    [ABCI++ specification](https://github.com/tendermint/tendermint/blob/4743a7ad0/spec/abci%2B%2B/README.md).
+
+    The catch-up mode does not need vote extension information to make progress, as all it needs is the
+    decided block at each height to call `FinalizeBlock` and keep the state-machine replication making
+    progress. The consensus mode, on the other hand, does need vote extension information at every height.
+
+    When a validator falls behind for whatever reason, e.g. cases (d), (e), (f), and (g) above, it
+    follows the catch-up mode as long as it considers itself out of date. The key safety property we need
+    to provide is the following:
+
+    - for every height *h*, a node in catch-up mode does not accept a (gossipped) block for height
+        *h+1* **until**:
+        - it has received and (light-client) verified all the blocks leading up to a new height
+          *h' > h*
+        - it has received an extended commit for *h'* and has verified:
+            - the precommit vote signatures in the extended commit
+            - the vote extension signatures in the extended commit (each being signed with the same
+              key as the precommit vote it extends)
+            - all non-`nil` precommit votes have their corresponding extension
+
+    This invariant ensures that the node in catch-up mode:
+
+    - cannot be attacked by being fed a block without extensions in a similar way as explained for
+      solution 2.
+    - at any moment, there exists a future height *h'* for which the node holds the (light-client
+      verified) block and *a* corresponding (verified) extended commit.
+
+    The second bullet above allows a node to have the *flexibility* to decide when to consider itself
+    as caught up (independently of any network conditions) and switch to consensus mode by using the
+    extended commit for *h'* to build a proposal in the rounds of *h' + 1* where it is the proposer.
+
+    Of course, the node may fall behind later on. In that case, it rejects switching back to consensus
+    mode until it receives a sequence of blocks and an extended commit that abides by the invariant
+    specified above.
+
+### Feasibility of the Proposed Solutions
+
+Solution 0, besides the drawbacks described in the previous section, provides guarantees that are
+weaker than the rest. The Application does not have the assurance that more than *2n/3* vote
+extensions will *always* be available when calling `PrepareProposal`.
+This level of guarantees is probably not strong enough for vote extensions to be useful for some of
+the use cases that motivated them in the first place, e.g., encrypted mempool transactions.
+
+Solution 1, while being simple in that the changes needed in the current Tendermint codebase would
+be rather small, is changing the block format, and would therefore require all blockchains using
+Tendermint v0.35 or earlier to hard-fork when upgrading to v0.36.
```

Comment:
This does make me somewhat concerned about how a chain can upgrade from v0.35->v0.36 in the current scheme since there are no vote extensions present and there won't be any when the node starts up. I'm wondering if we have ways around this? Potentially an initial height at which vote extensions will be expected :thinking:

Reply:
Good question. My understanding of a hard fork (the way it's been done in the past) is:
* agreeing on a "stop height"
* dumping the app's state at that height
* starting the chain with a different chain-ID using the dumped state as part of the genesis file

In a hard fork made like that, there is no problem, as height 1's block does not have a `LastCommit` (never has), and therefore it does not need vote extensions. If the App happens to _need_ vote extensions in height 1, that information should probably be included in the genesis file, along with the dumped state.

Reply:
I would agree with Sam that a hardfork for a feature that will only be used by _some_ chains is not viable. I would suggest that, as a way to allow this upgrade to proceed gracefully and without breaking all historical chains we gate it behind an application feature flag, probably a 'consensus parameter' to determine if these are required. This parameter, `RequireVoteExtensions` would default to `false`. Chains could therefore update the parameter to `true` after a few heights to enforce that the extensions exist and are useable.

Reply:
I came here to ask about this too :slightly_smiling_face: 

My question arose when looking at https://github.com/tendermint/tendermint/pull/8141#discussion_r848579379

Reply:
I like @williambanfield's feature gating solution. The only possible drawback I see here is that we'll need to thread this parameter through a whole lot of places in the code.

Reply:
Can we perhaps make the parameter an integer value that indicates a height? e.g. a config parameter like `expect_vote_extensions_from_height` or something?

Reply:
Absolutely. It needs a bit of wiring, but it's definitely doable. It's largely similar to: https://github.com/orgs/tendermint/projects/11

Reply:
The other alternative is just to detect when we start to see vote extension signatures from all validators and automatically consider that the starting point for vote extensions.

Reply:
@thanethomson Perhaps, although I'm not sure that's much preferred over a bool. There has to be a default that we set that works for at least a height: https://github.com/tendermint/tendermint/blob/master/docs/rfc/rfc-009-consensus-parameter-upgrades.md#allowing-developer-defined-values-and-the-endblock-problem

Reply:
For the record. I misunderstood @williambanfield's initial comment. I 100% agree we don't want to force chains to hard fork when upgrading to v0.36.
Let's discuss tomorrow what we can do.

Reply:
I think a consensus param for switching on and off vote extensions makes sense not just for the case of upgrading but also in the event that an application vulnerability is discovered which compromises the vote extension protocol. After all this is an experimental component.